pkg/controller/config/token.go

```diff
@@ -0,0 +1,180 @@
+		Complete(NewTokenReconciler(mgr, of,
+			WithLogger(l.WithValues("token-controller", name)),
+			WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+// A TokenReconciler reconciles managed resources by creating and managing the
+// lifecycle of an external resource, i.e. a resource in an external system such
+// as a cloud provider API. Each controller must watch the managed resource kind
+// for which it is responsible.
+type TokenReconciler struct {
+	client client.Client
+
+	newConfig func() resource.ProviderConfig
+
+	log    logging.Logger
+	record event.Recorder
+}
+
+// A TokenReconcilerOption configures a Reconciler.
+type TokenReconcilerOption func(*TokenReconciler)
+
+// WithLogger specifies how the Reconciler should log messages.
+func WithLogger(l logging.Logger) TokenReconcilerOption {
+	return func(r *TokenReconciler) {
+		r.log = l
+	}
+}
+
+// WithRecorder specifies how the Reconciler should record events.
+func WithRecorder(er event.Recorder) TokenReconcilerOption {
+	return func(r *TokenReconciler) {
+		r.record = er
+	}
+}
+
+// NewTokenReconciler returns a Reconciler of ProviderConfigs.
+func NewTokenReconciler(m manager.Manager, of resource.ProviderConfigKinds, o ...TokenReconcilerOption) *TokenReconciler {
+	nc := func() resource.ProviderConfig {
+		return resource.MustCreateObject(of.Config, m.GetScheme()).(resource.ProviderConfig)
+	}
+
+	// Panic early if we've been asked to reconcile a resource kind that has not
+	// been registered with our controller manager's scheme.
+	_ = nc()
+
+	r := &TokenReconciler{
+		client: m.GetClient(),
+
+		newConfig: nc,
+
+		log:    logging.NewNopLogger(),
+		record: event.NewNopRecorder(),
+	}
+
+	for _, ro := range o {
+		ro(r)
+	}
+
+	return r
+}
+
+// Reconcile a ProviderConfig by accounting for the managed resources that are
+// using it, and ensuring it cannot be deleted until it is no longer in use.
+func (r *TokenReconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("request", req)
+	log.Debug("Reconciling")
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	pc := &v1beta1.ProviderConfig{}
+	if err := r.client.Get(ctx, req.NamespacedName, pc); err != nil {
+		// In case object is not found, most likely the object was deleted and
+		// then disappeared while the event was in the processing queue. We
+		// don't need to take any action in that case.
+		log.Debug(errGetPC, "error", err)
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errGetPC)
+	}
+
+	ref := pc.Spec.Credentials.SecretRef
+	if ref == nil {
+		return reconcile.Result{}, errors.New(errNoSecret)
+	}
+
+	s := &v1.Secret{}
+	if err := r.client.Get(ctx, types.NamespacedName{Name: ref.Name, Namespace: ref.Namespace}, s); err != nil {
+		return reconcile.Result{}, errors.Wrap(err, errGetSecret)
+	}
+
+	config := bluemix.Config{
+		EndpointLocator: endpoints.NewEndpointLocator(getRegion(*pc)),
```

Comment:
This is where we use the Region to get the API endpoint for generating the bearer token used in all API calls.